### ISSUE TITLE
Stackdriver TimeSeries Points must be written in order

### ISSUE THREAD

**Issue by serathius** [kubernetes/heapster#1988]
Looks like heapster doesn't guarantee order of metrics sent

**Steps to reproduce:**
Run k8s cluster with larger number of nodes. In my case 200

**Describe the results you received:**
In 40% requests Stackdriver responds with 400
In logs visible:
```
W0323 09:34:18.012144       1 stackdriver.go:321] Error while sending request to Stackdriver googleapi: Error 400: One or more TimeSeries could not be written:
Points must be written in order. One or more of the points specified was older than the most recent stored point.: timeSeries[70-77], badRequest
```

**Describe the results you expected:**
requests to Stackdriver not being rejected

**heapster --version**
```
version: v1.4.3
commit: b1d70b4d
```

**kubectl version**
```
Client Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.12-gke.1", GitCommit:"f4f0bedb82a57dc32b21b8f15fe01e0b8411356b", GitTreeState:"clean", BuildDate:"2018-02-05T18:25:12Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.12-gke.1", GitCommit:"f4f0bedb82a57dc32b21b8f15fe01e0b8411356b", GitTreeState:"clean", BuildDate:"2018-02-05T18:21:59Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by kawych** [comment]
@x13n 
Daniel, IIRC you have been working on this bug some time ago - can you summarize here what was the outcome?

**Comment by x13n** [comment]
I didn't get to the root cause, unfortunately. My only finding was that in the request sent by Heapster, the container name was sometimes empty. At the same time, no such thing could be observer in the summary API. My guess was that there is a bug in Heapster causing it to sometimes lose container names, which in turn causes Stackdriver to think there are duplicates (if there was more than one container without a name set). I don't know if this is what is happening here and I don't know whether this is the only problem.

**Comment by kawych** [comment]
Thanks Daniel. Container name is empty for pod metrics, so if container metrics are reported with empty container name, this can lead to similar error even for one-container pods (not sure if the error message would be the same though).

**Comment by x13n** [comment]
I'm pretty sure that was cpu or memory usage - do we expose them as pod metrics in addition to exposing them as container metrics?

**Comment by kawych** [comment]
I'm not 100% sure. This component: https://github.com/kubernetes/heapster/blob/master/metrics/processors/pod_aggregator.go is responsible for aggregating all container metrics on pod level. It looks like it blacklists all Delta and Cumulative metrics. CPU usage is cumulative, so it should be blacklisted. Memory usage is gauge, so I think it's exposed also as pod metric.
